include/drivers/rtio.h

```diff
@@ -0,0 +1,487 @@
+ * into an rtio block.
+ */
+#define Z_RTIO_BLOCK_PUSH_PULL_IMPL(_type)				\
+	static inline int rtio_block_push_##_type(			\
+		struct rtio_block *block, _type val)			\
+	{								\
+		if (rtio_block_available(block) >= sizeof(_type)) {	\
+			memcpy(&block->_data[block->len], &val,		\
+			       sizeof(_type));				\
+			block->len += sizeof(_type);			\
+			return 0;					\
+		} else {						\
+			return -ENOMEM;					\
+		}							\
+	}								\
+									\
+	static inline int rtio_block_pull_##_type(			\
+		struct rtio_block *block, u16_t offset, _type * val)	\
+	{								\
+		if (block->len - offset >= sizeof(_type)) {		\
+			memcpy(val, &block->_data[offset],		\
+			       sizeof(_type));				\
+			return offset + sizeof(_type);			\
+		} else {						\
+			return -EINVAL;					\
+		}							\
+	}
+
+/**
+ * @brief Push/pull a value in/out of the block
+ *
+ * Implementations for common data types you might wish to
+ * store in a rtio_block
+ */
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(u8_t);
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(u16_t);
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(u32_t);
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(u64_t);
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(s8_t);
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(s16_t);
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(s32_t);
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(s64_t);
+
+struct rtio_block_allocator;
+
+/**
+ * @private
+ * @brief Function definition for allocating rtio blocks from a given allocator
+ */
+typedef int (*rtio_block_alloc_t)(struct rtio_block_allocator *allocator,
+				  struct rtio_block **block, u32_t size,
+				  u32_t timeout);
+
+/**
+ * @private
+ * @brief Function definition for freeing rtio blocks
+ */
+typedef void (*rtio_block_free_t)(struct rtio_block_allocator *allocator,
+				  struct rtio_block *block);
+
+/**
+ * @brief An rtio block allocator interface
+ */
+struct rtio_block_allocator {
+	rtio_block_alloc_t alloc;
+	rtio_block_free_t free;
+};
+
+/**
+ * @brief An rtio mempool block allocator
+ */
+struct rtio_mempool_block_allocator {
+	struct rtio_block_allocator allocator;
+	struct k_mem_pool *mempool;
+};
+
+/**
+ * @brief An rtio_block allocated from a mempool
+ */
+struct rtio_mempool_block {
+	struct rtio_block block;
+	struct k_mem_block_id id;
+};
+
+/**
+ * @brief Allocate a block from a mempool
+ */
+static inline int rtio_mempool_block_alloc(
+	struct rtio_block_allocator *allocator, struct rtio_block **block,
+	u32_t size, u32_t timeout)
+{
+	struct rtio_mempool_block_allocator *mempool_allocator =
+		(struct rtio_mempool_block_allocator *)allocator;
+	struct k_mem_block memblock;
+	size_t block_size = size + sizeof(struct rtio_mempool_block);
+
+	int res = k_mem_pool_alloc(mempool_allocator->mempool,
+				   &memblock, block_size,
+				   timeout);
+	if (res == 0) {
+		struct rtio_mempool_block *mempool_block =
+			(struct rtio_mempool_block *)(memblock.data);
+		size_t struct_size = sizeof(struct rtio_block);
+		u8_t *dataptr = &((u8_t *)memblock.data)[struct_size];
```

Comment:
@bfrog Given your comment on how the buffer is not contiguous with the memory immediately after `struct rtio_block` the calculation here looks wrong.  Because it looks like that's what's being stored.  Which overlaps with `mempool_block->id`.  I think.

Maybe the `struct_size` above should be `sizeof(struct rtio_mempool_block)` then.  Which I'd express as:

    u8_t *data_ptr = (u8_t *)memblock.data + sizeof(*mempool_block);

Reply:
yes, thank you, this is an error when switching to allowing any allocator

Reply:
Fixed in the latest